modin/core/dataframe/pandas/dataframe/dataframe.py

```diff
@@ -4629,3 +4636,179 @@ def from_dataframe(cls, df: "ProtocolDataframe") -> "PandasDataframe":
+        single_wrap = Engine.get() != "Dask"
+        cls = type(self)
+        wrapper_put = self._partition_mgr_cls._execution_wrapper.put
+        if (
+            not single_wrap
+            or (remote_fn := getattr(cls, "_CASE_WHEN_FN", None)) is None
+        ):
+
+            def case_when(df, caselist):  # pragma: no cover
+                caselist = [
+                    tuple(
+                        data.iloc[:, 0] if isinstance(data, pandas.DataFrame) else data
+                        for data in case_tuple
+                    )
+                    for case_tuple in caselist
+                ]
+                series = df.iloc[:, 0]
+                return pandas.DataFrame({series.name: series.case_when(caselist)})
+
+            if single_wrap:
+                cls._CASE_WHEN_FN = remote_fn = wrapper_put(case_when)
+            else:
+                remote_fn = case_when
+
+        parts_len = len(self._partitions)
+        parts = None
+        lengths = None
+        is_trivial_idx = None
+
+        def copartition(df, fill_value):
+            nonlocal is_trivial_idx
+            copartition = False
+            if is_trivial_idx is None:
+                is_trivial_idx = is_trivial_index(self.index)
+            if (
+                is_trivial_idx != is_trivial_index(df.index)
+                or (not is_trivial_idx and any(self.index != df.index))
+                or (parts_len < len(df._partitions))
+            ):
+                copartition = True
+            else:
+                nonlocal parts, lengths
+                if parts is None:
+                    parts = [p[0] for p in self._partitions]
+                    lengths = self._get_lengths(parts, Axis.ROW_WISE)
+
+                df_parts = [p[0] for p in df._partitions]
+                df_part_lengths = df._get_lengths(df_parts, Axis.ROW_WISE)
+                if any(lengths[i] != df_part_lengths[i] for i in range(len(df_parts))):
+                    copartition = True
+            if copartition:
+                _, list_of_right_parts, joined_index, row_lengths = self._copartition(
+                    Axis.ROW_WISE.value,
+                    df,
+                    how="left",
+                    sort=False,
+                    fill_value=fill_value,
+                )
+                df = self.__constructor__(
+                    list_of_right_parts[0],
+                    joined_index,
+                    df.columns,
+                    row_lengths,
+                    df.column_widths,
+                )
+            return df
+
+        use_map = single_wrap
+        new_caselist = []
+        for condition, replacement in caselist:
+            if callable(condition):
+                if single_wrap:
+                    condition = wrapper_put(condition)
+            else:
+                use_map = False
+                if isinstance(condition, cls):
+                    condition = copartition(condition, True)
```

Comment:
btw, `._copartition()` support copartitioning along several dataframes at the same time, would it be more efficient to first filter out all modin dataframes and then call `._copartition()` only once?